speechbrain/lobes/models/huggingface_transformers/huggingface_new.py

```diff
@@ -0,0 +1,424 @@
+    """
+
+    def __init__(
+        self,
+        source,
+        save_path="",
+        for_pretraining=False,
+        with_lm_head=False,
+        with_casual_lm=False,
+        freeze=False,
+        cache_dir="pretrained_models",
+        **kwarg,
+    ):
+        super().__init__()
+
+        # Fetch config
+        self.config, _unused_kwargs = AutoConfig.from_pretrained(
+            source, cache_dir=save_path, return_unused_kwargs=True,
+        )
+
+        self.config = self.override_config(self.config)
+
+        self.for_pretraining = for_pretraining
+        self.with_lm_head = with_lm_head
+        if for_pretraining:
+            model = AutoModelForPreTraining.from_config(self.config)
+        elif with_lm_head:
+            model = AutoModelWithLMHead.from_config(self.config)
+        elif with_casual_lm:
+            model = AutoModelForCausalLM.from_config(self.config)
+        else:
+            model = AutoModel.from_config(self.config)
+
+        
+        # Download model
+        if "quantization_config" in kwarg:
+            bnb_conifg= kwargs["quantization_config"]
+        else:
+            bnb_conifg= None
+        self._from_pretrained(
+            source,
+            config=self.config,
+            model=model,
+            save_path=save_path,
+            cache_dir=cache_dir,
+            quantization_config=bnb_config
+        )
+
+        # Prepare for training, fine-tuning, or inference
+        self.freeze = freeze
+        if self.freeze:
+            logger.warning(
+                f"speechbrain.lobes.models.huggingface_transformers.huggingface - {type(self.model).__name__} is frozen."
+            )
+            self.freeze_model(self.model)
+        else:
+            self.model.gradient_checkpointing_disable()  # Required by DDP
+            self.model.train()
+
+    def _from_pretrained(
+        self, source, config, model, save_path, cache_dir,quantization_config=None
+    ):
+        """This function manages the source checking and loading of the params.
+
+        # 1. Is the model from HF or a local path
+        # 2. Is the model pretrained with HF or SpeechBrain
+        # 3. Download (if appropriate) and load with respect to 1. and 2.
+
+        Arguments
+        ---------
+        source : str
+            HuggingFace hub name: e.g "facebook/wav2vec2-large-lv60"
+        config : AutoConfig
+            HuggingFace generic configuration class.
+        model: AutoModel
+            HuggingFace generic model class.
+        save_path : str
+            Path (dir) of the downloaded model.
+        cache_dir : str
+            Path (dir) in which a downloaded pretrained model configuration should be cached.
+        """
+        is_sb, ckpt_file, is_local = self._check_model_source(source, save_path)
+        if is_sb:
+            config = config.from_pretrained(source, cache_dir=save_path,quantization_config=quantization_config)
```

Comment:
@poonehmousavi : Do we also need quantization_config in the case when is_sb = False, i,e, adding quantization_config in the below else condition?